rwx-research/abq
# Prevent scenario where PAT is used alongside nonexistent run_id

We detect a PAT and no run ID was provided to the CLI -> ephemeral

We detect a PAT and a run ID provided, but doesn't exist (as detected by the server) -> error

- [x] Requires a rwx cloud server change be made prior to merge.

## QA Plan

<details>
<summary>
✅ Observe using a PAT with an explicit, but not existing run id, results in an error.
</summary>

```
% target/debug/abq test -n 1 --run-id tonytest3 -- bundle exec rspec bigtest/benchmark_rspec/benchmark_spec.rb
Error: error: ABQ was unable to find a queue to run against. Personal access tokens require an existing run_id

Usage: abq <COMMAND>

For more information, try '--help'.
```

</details>

<details>
<summary>
✅  Observe using a PAT with no run id uses an ephemeral queue
</summary>

```
Generic test runner for a24df93c-469f-4768-a42a-0e8e2f738c05 started on worker af214cc2-bd2f-49c1-a133-21c3b4abc924 (worker 0, runner 1)
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 18.8 seconds (files took 0.07626 seconds to load)
500 examples, 0 failures



--------------------------------------------------------------------------------
------------------------------------- ABQ --------------------------------------
--------------------------------------------------------------------------------

Finished in 19.56 seconds (18.09 seconds spent in test code)
500 tests, 0 failures
```

</details>

<details>
<summary>
✅ Org token happily starts a remote test run w/ a non-existing run id
</summary>

```
% target/debug/abq test -n 1 -- bundle exec rspec bigtest/benchmark_rspec/benchmark_spec.rb
Generic test runner for f635bf93-045b-4c8e-8a7f-ecce8f7437d2 started on worker 3280f25f-ee3c-4d5a-bcca-d6ebac2fcfee (worker 0, runner 1)
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 18.91 seconds (files took 0.23486 seconds to load)
500 examples, 0 failures



--------------------------------------------------------------------------------
------------------------------------- ABQ --------------------------------------
--------------------------------------------------------------------------------

Finished in 20.02 seconds (18.14 seconds spent in test code)
500 tests, 0 failures


Run the following command to replay these tests locally:


        abq test --run-id f635bf93-045b-4c8e-8a7f-ecce8f7437d2 --worker 0 --num 1 -- <your-test-command>


Specify your Access Token with the RWX_ACCESS_TOKEN env variable, passing --access-token, or running `abq login`.
```
</details>

<details>
<summary>
✅ Org token uses ephemeral queue when not given a run id


</summary>

```
% target/debug/abq test -n 1 --run-id tonytest4 -- bundle exec rspec bigtest/benchmark_rspec/benchmark_spec.rb
Generic test runner for tonytest4 started on worker a1e5e788-28f9-44a4-b5bd-92277821031b (worker 0, runner 1)
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 18.91 seconds (files took 0.25534 seconds to load)
500 examples, 0 failures



--------------------------------------------------------------------------------
------------------------------------- ABQ --------------------------------------
--------------------------------------------------------------------------------

Finished in 20.01 seconds (18.11 seconds spent in test code)
500 tests, 0 failures


Run the following command to replay these tests locally:


        abq test --run-id tonytest4 --worker 0 --num 1 -- <your-test-command>


Specify your Access Token with the RWX_ACCESS_TOKEN env variable, passing --access-token, or running `abq login`.
```

^ Not 100% sure how to confirm visually whether this was on an ephemeral queue 😅 
</details>
